### PR TITLE
Decimal support for add and subtract

### PR DESCRIPTION
**Pull request by rwlee**
Combines work from @sperlingxx #976 and @revans2 changes to resolve precision mismatches when running on the GPU. Merging this should close #976, as the necessary bits have already been merged or were copied over and modified for this PR.

### CODE REVIEW

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala [hunk 1]
```diff
@@ -16,10 +16,95 @@
...
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: DataFromReplacementRule)
+  extends UnaryExprMeta[PromotePrecision](expr, conf, parent, rule) {
+  override def convertToGpu(child: Expression): GpuExpression =
+    GpuPromotePrecision(child)
+}
+
+object DecimalExpressions {
+  // Underlying storage type of decimal data in cuDF is int64_t, whose max capacity is 19.
+  val GPU_MAX_PRECISION: Int = 19
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala:68]: Despite this code never actually being used. CUDF thinks this value should be 18

https://github.com/rapidsai/cudf/blob/639049808913c7d5a7170141d784da6cc45da36d/java/src/main/java/ai/rapids/cudf/DType.java#L25

So which is right? and why does this not just pull from that value?

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala [hunk 2]
```diff
@@ -16,10 +16,95 @@
...
+    expr: PromotePrecision,
+    conf: RapidsConf,
+    parent: Option[RapidsMeta[_, _, _]],
+    rule: DataFromReplacementRule)
+  extends UnaryExprMeta[PromotePrecision](expr, conf, parent, rule) {
+  override def convertToGpu(child: Expression): GpuExpression =
+    GpuPromotePrecision(child)
+}
+
+object DecimalExpressions {
```
Comment (revans2) [inline sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala:66]: This entire object is never used.  Why do we need it?

File: tests/src/test/scala/com/nvidia/spark/rapids/unit/DecimalUnitTest.scala [hunk 3]
```diff
@@ -21,8 +21,9 @@ import java.math.RoundingMode
 import scala.util.Random
 
 import ai.rapids.cudf.{ColumnVector, DType, HostColumnVector}
-import com.nvidia.spark.rapids.{GpuAlias, GpuBatchScanExec, GpuColumnVector, GpuIsNotNull, GpuIsNull, GpuLiteral, GpuOverrides, GpuScalar, GpuUnitTests, HostColumnarToGpu, RapidsConf}
+import com.nvidia.spark.rapids._
 
+import org.scalatest.Matchers.convertToAnyShouldWrapper
 import org.apache.spark.SparkConf
```
Comment (revans2) [inline tests/src/test/scala/com/nvidia/spark/rapids/unit/DecimalUnitTest.scala:27]: Could we totally revert all of the changes to this file, as it looks like they are not needed.

**Review by revans2 — COMMENTED** [review]
Because I was involved with writing some of the parts of the PR I think it is best if others approve/review it too.

File: sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala (hunk 2 above)
Comment (rwlee) [reply sql-plugin/src/main/scala/com/nvidia/spark/rapids/decimalExpressions.scala:66]: I don't think it's needed anymore, all references were stripped out with the simplification of the `PromotePrecision` operation.

**Review by revans2 — APPROVED** [review]

**Comment by revans2** [comment]
build

**Review by jlowe — APPROVED** [review]

**Pull request merged**